presto-main/src/main/java/com/facebook/presto/operator/WorkProcessorUtils.java

```diff
@@ -162,42 +162,31 @@ else if (processor.isBlocked()) {
     {
         requireNonNull(processor, "processor is null");
         requireNonNull(transformation, "transformation is null");
-        return processor.transform(new Transformation<T, R>()
-        {
-            WorkProcessor<R> processor;
-            boolean needsMoreData;
-
-            @Override
-            public ProcessorState<R> process(Optional<T> elementOptional)
-            {
-                while (true) {
-                    if (processor == null) {
-                        ProcessorState<WorkProcessor<R>> state = transformation.process(elementOptional);
-                        if (state.getType() != ProcessorState.Type.RESULT) {
-                            return new ProcessorState<>(state.getType(), state.isNeedsMoreData(), Optional.empty(), state.getBlocked());
-                        }
-                        processor = state.getResult().get();
-                        needsMoreData = state.isNeedsMoreData();
-                    }
+        return flatten(processor.transform(transformation));
+    }
 
-                    if (processor.process()) {
-                        if (!processor.isFinished()) {
-                            return ProcessorState.ofResult(processor.getResult(), false);
-                        }
+    static <T> WorkProcessor<T> flatten(WorkProcessor<WorkProcessor<T>> processor)
+    {
+        requireNonNull(processor, "processor is null");
+        return processor.transform(nestedProcessorOptional -> {
+            if (!nestedProcessorOptional.isPresent()) {
+                return ProcessorState.finished();
+            }
 
-                        processor = null;
-                        if (needsMoreData) {
-                            return ProcessorState.needsMoreData();
-                        }
-                    }
-                    else if (processor.isBlocked()) {
-                        return ProcessorState.blocked(processor.getBlockedFuture());
-                    }
-                    else {
-                        return ProcessorState.yield();
-                    }
+            WorkProcessor<T> nestedProcessor = nestedProcessorOptional.get();
+            if (nestedProcessor.process()) {
+                if (nestedProcessor.isFinished()) {
+                    return ProcessorState.needsMoreData();
```

Comment:
Why? The nested processor didn't need more data. We 
Add javadocs in ProcessorState and/or Transformation and/or `com.facebook.presto.operator.WorkProcessor#transform`